components/inputs/input-date-range.js

```diff
@@ -129,16 +149,71 @@ class InputDateRange extends RtlMixin(LocalizeCoreElement(LitElement)) {
 		if (input) input.focus();
 	}
 
+	async validate(showErrors) {
+		const errors = await super.validate(showErrors);
+		if (errors.length !== 0) {
+			return errors;
+		}
+		setTimeout(async() => {
+			return Promise.all([
+				await this.shadowRoot.querySelector('.d2l-input-date-range-start').validate(showErrors),
+				await this.shadowRoot.querySelector('.d2l-input-date-range-end').validate(showErrors)]
+			).then((res) => {
+				return res.reduce((acc, errors) => [...acc, ...errors], []);
+			});
+		});
+	}
+
+	get validationMessageBadInput() {
+		return this.localize('components.input-date-range.errorBadInput', { startLabel: this._startLabel, endLabel: this._endLabel });
+	}
+
+	get _endLabel() {
+		return this.endLabel ? this.endLabel : this.localize('components.input-date-range.endDate');
+	}
+
 	async _handleChange(e) {
-		// TODO: validation that start is before end
 		const elem = e.target;
-		if (elem.classList.contains('d2l-input-date-range-start')) this.startValue = elem.value;
-		else this.endValue = elem.value;
+		if (elem.classList.contains('d2l-input-date-range-start')) {
+			this.startValue = elem.value;
+		} else {
+			this.endValue = elem.value;
+		}
+		await this.requestValidate();
 		this.dispatchEvent(new CustomEvent(
 			'change',
 			{ bubbles: true, composed: false }
 		));
 	}
 
+	_handleDropdownOpen(e) {
+		const id = e.target.id;
+		const tooltip = this.shadowRoot.querySelector(`d2l-tooltip[for="${id}"]`);
+		if (tooltip && tooltip.showing) tooltip.hide();
+	}
+
+	_handleInnerFocus(e) {
+		// in order for calendar close to cause tooltip to show
+		const id = e.target.id;
+		const tooltip = this.shadowRoot.querySelector(`d2l-tooltip[for="${id}"]`);
+		if (tooltip && !tooltip.showing) tooltip.show();
+	}
+
+	_handleNestedFormElementValidation(e) {
+		e.preventDefault();
+	}
+
+	get _startLabel() {
+		return this.startLabel ? this.startLabel : this.localize('components.input-date-range.startDate');
+	}
```

Comment:
I assume this isn't inlined in render for performance reasons? If the localize mixin doesn't automatically cache terms is this something we should consider adding? It feels a bit weird to have components be responsible for caching their own lang terms.

Reply:
It was suggested here https://github.com/BrightspaceUI/core/pull/693#discussion_r455058275, more for simplicity than performance. I think localize mixin does some sort of caching of the lang terms but I haven't been in that code for a bit so I don't exactly recall - will take a look.

Reply:
I'm fine leaving it as is if it benefits simplicity. But it would be nice to remove the `startLabel` property if `localize` is caching it for us so we can just `return this.localize('components.input-date-range.startDate');`

Reply:
So `startLabel` is for if the consumer of `input-date-range` wants to specify what the start label is (e.g., instead of "Start Date" it's "Begin on.." or something like that) whereas `_startLabel` is the internal usage of that, so I'm thinking I should rename `_startLabel`

Reply:
Sorry about that, missed that there was a corresponding public `startLabel` property for users of the component. That makes sense to me!